Initial Intel HEXL integration

fboemer opened the pull request:
Initial integration with Intel HEXL (https://github.com/intel/hexl)

Co-authored-by: Gelila Seifu <gelila.seifu@intel.com>
Co-authored-by: Jeremy Bottleson <jeremy.bottleson@intel.com>

fionser commented:
@fboemer  Great boost. HEXL saves ~50% computation time for my program :).
But, my project uses SEAL as a submodulue and it failed to build with HEXL in a submodulue.
I need to build & install SEAL seperately, and then build my own project.

WeiDaiWD commented:
> Great boost. HEXL saves ~50% computation time for my program :).
Can you share your CPU spec? Does it have `AVX512IFMA`?

fionser commented:
- Intel(R) Xeon(R) Platinum 8269CY CPU @ 2.50GHz
- Does not have AVX512IFMA
- gcc version 7.2.1 20180104

CPU-Flags: 

```
fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc eagerfpu pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch arat invpcid_single fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f avx512dq rdseed adx smap avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1
```

WeiDaiWD commented:
Cool! It's remarkably faster even with `avx512dq`.

WeiDaiWD commented:
@fboemer This is superb. Thanks for making SEAL faster. :)

A few things to add to this PR:
- `reinterpret_cast<const uint64_t *>` and `reinterpret_cast<uint64_t *>` in `polyarithsmallmod.h` and `polyarithsmallmod.cpp` are not necessary, since we defined `using CoeffIter = PtrIter<std::uint64_t *>` in `iterator.h`.
- Two CMake options, `BUILD_PIC` and `BUILD_TESTING`, are propagated into SEAL from Intel HEXL. It's not obvious to me how they end up there by eyeballing HEXL's `CMakeLists.txt`. Could you please figure it out and remove or hide them?
- Could you please provide descriptions (of functionality and important arguments) to the three functions in `intel_seal_ext.h`? 

Questions:
- `SEAL_ALIGNED_ALLOC` calls `malloc` if size is not a multiple of alignment. What's the effect on performance here? My only worry is that a user can be falsely convinced that all allocations are aligned to 64 as long as `SEAL_USE_ALIGN_64` is `ON`.
- In HEXL's README, it says "Intel HEXL targets integer arithmetic with word-sized primes, typically 40-60 bits." BFV auxiliary primes are 61-bit. Does AVX512IFMA only work on 54-bit or less? What will happen if a prime is larger than the bound (60 or 54)?

Recommendations to Intel HEXL:
- As fionser said, would it make sense for Intel HEXL to use the powers of root (specifically their allocation) generated by SEAL, or disable SEAL's NTT precomputation if `SEAL_USE_INTEL_HEXL=ON`? The former choice removes precomputation in the first call to `intel::seal_ext::get_ntt`.
- When compiling your branch in debug mode (several warnings enabled), I've got [a long list of warnings](https://gist.github.com/WeiDaiWD/55e7dd614fb9a231b6b82c7f745ea6c1). They are categorized into followings:
    - sign conversion
    - implicit int float conversion
    - shorten 64 to 32
    - c++14 binary literal: maybe Intel HEXL should use C++14 instead of C++11

fboemer commented:
Thanks for the feedback, @WeiDaiWD  and @fionser.

A few notes:
- @fionser , do you mind trying the submodule approach again and reporting the errors? We're happy to support this workflow if possible
- We've removed the unnecessary `reinterpret_cast`s. Thanks for pointing this out; it's a clean approach!
- `BUILD_PIC` and `BUILD_TESTING` should no longer be leaking (they were stemming from some 3rd-party dependenceis)
- Within Intel HEXL, I've measured the 64-byte aligned allocations to yield ~5-7% speedup on the NTT. It looks like whenever we allocate memory for cipher/plaintexts, the allocation sizes will be a multiple of 64 bytes. So the current implementation should be efficient. Otherwise, for other-sized memory allocations, we could use an approach similar to [Boost](https://www.boost.org/doc/libs/1_75_0/boost/align/detail/aligned_alloc.hpp), which will allocate extra memory.
- We've added documentation for the `intel_seal_ext.h`
- AVX512IFMA52 performs 52-bit integer arithmetic. We need a few extra bits in the NTT, so choosing coefficient moduli < 50 bits should suffice for best performance. For large primes, e.g. the auxiliary prime, HEXL will choose an AVX512DQ implementation, which still yields some speedup (as @fionser  observed), but less than the IFMA52 approach. See Tables 1-4 in our [arXiv paper](https://arxiv.org/pdf/2103.16400.pdf) for more information
- Regarding NTT pre-computation: Intel HEXL uses two forms of pre-computation. One is based on Barrett factors `floor(2^64/modulus)` and today happens to have the same order as SEAL's pre-computation. A second pre-computation vector is based on `floor(2^52/modulus)`. In general, the pre-computation is not part of HEXL's public API. If it changes down the road, we don't want to break the SEAL integration. Yes, we could omit the SEAL NTT tables pre-computation if you like. We thought the current implementation would be safest, so as not to break any programs that may rely on SEAL's pre-computed NTT tables.
- Thanks for pointing our the warnings. We've updated Intel HEXL v1.0.0, which should have resolved them. Just a note that we plan to update the HEXL v1.0.0 tag a few more times for minor README changes, but will freeze once this PR is approved.

Also a note that this updated PR makes two more changes:
- Run pre-commit  on all the files. Maybe my system is different, but this lead to a few minor changes
- Adds STATUS messages for the CMake options. This makes it easier to tell what options are enabled when building from command line.

WeiDaiWD commented:
@fionser Does your program use BGV or BFV?

WeiDaiWD commented:
@fboemer Everything looks good now. Would you let me know when you think 1.0.0 is stable enough to freeze the tag? I'm ready to merge this into SEAL at any time, then why not wait for your upcoming commit hashes.

Side note:
I did some experiment with BFV. I set prime bit limits to 49 and 50 and deleted two default parameters to avoid errors. Compared to your branch, `BFV / EvaluateMultCt` gets faster in [my branch](https://github.com/WeiDaiWD/SEAL/tree/wei/test_bfv) for larger parameters. I'm testing on a Intel® Xeon(R) Silver 4108 @ 1.80 GHz, GCC 7.5.0. You should try on an IceLake processor. If it helps to get BFV faster, I can try to make those prime size limits configurable easier (without errors).

fboemer commented:
@WeiDaiWD , we updated the tag to v1.0.1 after all. No more changes will be made to this tag, so feel free to do any final testing and merge.

Thanks for trying out the prime bit-width changes. I'll try them out on an IceLake processor and report my findings here.

WeiDaiWD commented:
Cool. I'll just merge this. Let me know if you see important to update the tag/commit before the next SEAL's release. Thanks!

WeiDaiWD commented:
One little suggestion: when I build it on a Core i7-10700K that does not have AVX512, I have the following warning:
`warning: AVX512F vector return without AVX512F enabled changes the ABI [-Wpsabi]`

Maybe you want to explain to users what it means.

fionser commented:
> @fionser Does your program use BGV or BFV?

CKKS only

fionser commented:
@WeiDaiWD @fboemer 

Interestingly, when used HEXL, the BFV decryption slow down quite significantly.
```
/
| Encryption parameters :
|   scheme: BFV
|   poly_modulus_degree: 4096
|   #moduli: 2
|   #special_primes: 1
|   coeff_modulus size: 109 (59 + 50) bits
|   plain_modulus: 4194304 (23) bits
\
```
BFV decryption took 3.1ms/ 0.4ms (w and w/o HEXL) in my machine (gcc version 7.2.1, Red Hat 7.2.0-5, Intel(R) Xeon(R) Platinum 8269CY CPU @ 2.50GHz)

fionser commented:
@fboemer Nice thank you for the information.
